torchmultimodal/modules/encoders/mdetr_text_encoder.py

```diff
@@ -0,0 +1,240 @@
+
+        if position_ids is None:
+            # Create the position ids from the input token ids. Any padded tokens remain padded.
+            position_ids = create_position_ids_from_input_ids(
+                input_ids, self.padding_idx
+            )
+
+        if attention_mask is None:
+            attention_mask = torch.ones(((batch_size, seq_length)), device=device)
+
+        if token_type_ids is None:
+            if hasattr(self, "token_type_ids"):
+                buffered_token_type_ids = self.token_type_ids[:, :seq_length]  # type: ignore
+                buffered_token_type_ids_expanded = buffered_token_type_ids.expand(
+                    batch_size, seq_length
+                )
+                token_type_ids = buffered_token_type_ids_expanded
+            else:
+                token_type_ids = torch.zeros(
+                    batch_size,
+                    seq_length,
+                    dtype=torch.long,
+                    device=device,
+                )
+
+        if inputs_embeds is None:
+            inputs_embeds = self.word_embeddings(input_ids)
+        token_type_embeddings = self.token_type_embeddings(token_type_ids)
+
+        embeddings = inputs_embeds + token_type_embeddings
+        position_embeddings = self.position_embeddings(position_ids)
+        embeddings += position_embeddings
+
+        embeddings = self.LayerNorm(embeddings)
+        embeddings = self.dropout(embeddings)
+        return embeddings
+
+
+# Wrap TorchText's TransformerEncoder to take in embeddings instead of tokens
+class WrappedTransformerEncoder(nn.Module):
+    def __init__(
+        self,
+        embedding_dim: int,
+        num_encoder_layers: int,
+        num_attention_heads: int,
+        ffn_dimension: Optional[int] = None,
+        dropout: float = 0.1,
+        normalize_before: bool = False,
+        return_all_layers: bool = False,
+    ):
+        super().__init__()
+        ffn_dimension = ffn_dimension or 4 * embedding_dim
+        layer = torch.nn.TransformerEncoderLayer(
+            d_model=embedding_dim,
+            nhead=num_attention_heads,
+            dim_feedforward=ffn_dimension,
+            dropout=dropout,
+            activation="gelu",
+            batch_first=True,
+            norm_first=normalize_before,
+        )
+        self.layers = torch.nn.TransformerEncoder(
+            encoder_layer=layer, num_layers=num_encoder_layers
+        )
+        self.normalize_before = normalize_before
+        self.embedding_layer_norm = nn.LayerNorm(embedding_dim)
+        self.return_all_layers = return_all_layers
+        self.embedding_dim = embedding_dim
+
+    def _forward_return_all_layers(
+        self,
+        embeddings: torch.Tensor,
+        attn_mask: Optional[torch.Tensor] = None,
+    ) -> List[torch.Tensor]:
+        encoded = embeddings
+        states = [encoded]
+        for layer in self.layers.layers:
+            encoded = layer(encoded, src_key_padding_mask=attn_mask)
```

Comment:
@abhinavarora @parmeet this may be the source of the discrepancy

Reply:
@ebsmothers I am not sure what attn_mask is referring to here? could you provide pointer to the use site to understand what is being passed in the forward function. (Ideally this should be the mask which indicates which tokens are padding tokens i.e this mask is true where tokens match Padding Index, otherwise false)

Could you please help understand why this is a source of discrepancy? 

BTW I just compared the torchtext's output with fairseq's output at padded tokens, and it matches.

Reply:
@parmeet you can see the usage [here](https://github.com/ashkamath/mdetr/blob/main/models/transformer.py#L120-L121). Here `tokenized` contains `attention_mask`, which has shape `(batch_size, seq_len)` and tells us which tokens are padded. The problem is when we use TorchText's RoBERTaEncoder, we cannot modify this field at all. 

Starting from the bottom of the stack in `nn.MultiheadAttention`, you can see [here](https://github.com/pytorch/pytorch/blob/38e717dc876e0f9af66b32e9cc59b44b1f9edf27/torch/nn/modules/activation.py#L1024-L1032) that `attn_mask` there has shape `(target_seq_len, source_seq_len)`, which we do not want. Instead, we want `key_padding_mask`, which has shape `(batch_size, seq_len)`.

These two arguments are passed from `src_mask` and `src_key_padding_mask`, respectively, in `nn.TransformerEncoderLayer` ([ref](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/transformer.py#L460-L474)). But you can see that in TorchText's `TransformerEncoder` the value passed to `src_key_padding_mask` is constructed in the `forward()` ([ref](https://github.com/pytorch/text/blob/main/torchtext/models/roberta/modules.py#L139-L160)). In other words, we cannot actually pass the attention mask(padding mask? sorry, it's very hard to be consistent with naming here...) we get back from the tokenizer!

This is why I pointed out this line btw, as you can see in the [equivalent line](https://github.com/pytorch/text/blob/main/torchtext/models/roberta/modules.py#L160) in TorchText the value passed to `src_key_padding_mask` is not exposed in the API.

Hope this helps clarify things! Please let me know if anything is still unclear.

Reply:
Yes @ebsmothers, what you said is right. But why do we need to expose the padding mask? It is automatically constructed internally and is being passed to torch.nn.TransformerEncoderLayer. Sorry for noob question, but I am still not sure why this is source of discrepancy?  Assuming we expose the paddding mask in the API, what value would you be passing here to match the results from HF?

Reply:
@parmeet the padding mask seems to be used differently in the HF implementation than in PyTorch. I am not sure which is the "correct" way per se, but the implementation here (messy though it may be) is the only way I have been able to reproduce the results from the MDETR paper. There are two main differences between this version and the one in TorchText:

(1) inversion of the attention mask before passing it to TransformerEncoder
(2) my version doesn't apply the mask at the top level of the encoder 

Without both of these changes, I cannot match the results from MDETR. I think (1) is pretty standard, I know that HF and PyTorch have different mask definitions. But (2) is a bit more puzzling to me. Looking at the HF RoBERTa implementation, I do not see any use of their `attention_mask` outside of the self-attention layers. The `TransformerEncoder` layer in PyTorch core is similar. But in the TT `TransformerEncoder` for RoBERTa I see [this line](https://github.com/pytorch/text/blob/ed44dc88539b51785df03a276410cf71adfe6bfd/torchtext/models/roberta/modules.py#L309) applying the mask before passing to the encoder layers. Is this part of the standard RoBERTa implementation? If so, I am wondering why it's missing from the HF version.

Reply:
Thanks @ebsmothers for details. Regarding 2, It's a good question, I am not sure why HF version do not account for padding before passing to the encoder layers. I guess this explains the difference why the results mismatch (though only at padded tokens). The Fairseq version seems to be taking this into account [here](https://github.com/facebookresearch/fairseq/blob/08fe88479f8dd1f53705e203474a031d14cc2d75/fairseq/models/transformer/transformer_encoder.py#L211).